_studio/mfx_lib/shared/src/libmfxsw_decode.cpp

```diff
@@ -491,74 +491,88 @@ mfxStatus MFXVideoDECODE_DecodeFrameAsync(mfxSession session, mfxBitstream *bs,
-            // need only because SW HEVC decode is pseudo
+            // source data is OK, go forward
+            if (task.entryPoint.pRoutine)
             {
-                mfxPlugin plugin;
-                mfxPluginParam par;
-                if (session->m_plgDec.get())
+                mfxStatus mfxAddRes;
+
+                task.pOwner = session->m_pDECODE.get();
+                task.priority = session->m_priority;
+                task.threadingPolicy = session->m_pDECODE->GetThreadingPolicy();
+                // fill dependencies
+                task.pSrc[0] = *surface_out;
+                task.pDst[0] = *surface_out;
+                // this is wa to remove external task dependency for HEVC SW decode plugin.
+                // need only because SW HEVC decode is pseudo
                 {
-                    session->m_plgDec.get()->GetPlugin(plugin);
-                    MFX_CHECK_STS(plugin.GetPluginParam(plugin.pthis, &par));
-                    if (MFX_PLUGINID_HEVCD_SW == par.PluginUID)
+                    mfxPlugin plugin;
+                    mfxPluginParam par;
+                    if (session->m_plgDec)
                     {
-                        task.pDst[0] = 0;
+                        session->m_plgDec->GetPlugin(plugin);
+                        MFX_CHECK_STS(plugin.GetPluginParam(plugin.pthis, &par));
+                        if (MFX_PLUGINID_HEVCD_SW == par.PluginUID)
+                        {
+                            task.pDst[0] = 0;
+                        }
                     }
                 }
+
+    #ifdef MFX_TRACE_ENABLE
+                task.nParentId = MFX_AUTO_TRACE_GETID();
+                task.nTaskId = MFX::CreateUniqId() + MFX_TRACE_ID_DECODE;
+    #endif
+
+                // register input and call the task
+                mfxAddRes = session->m_pScheduler->AddTask(task, &syncPoint);
+                MFX_CHECK_STS(mfxAddRes);
             }
 
-#ifdef MFX_TRACE_ENABLE
-            task.nParentId = MFX_AUTO_TRACE_GETID();
-            task.nTaskId = MFX::CreateUniqId() + MFX_TRACE_ID_DECODE;
-#endif
+            if (MFX_ERR_MORE_DATA_SUBMIT_TASK == static_cast<int>(mfxRes))
+            {
+                mfxRes = MFX_WRN_DEVICE_BUSY;
+            }
 
-            // register input and call the task
-            mfxAddRes = session->m_pScheduler->AddTask(task, &syncPoint);
-            MFX_CHECK_STS(mfxAddRes);
-        }
+            if (MFX_WRN_DEVICE_BUSY == mfxRes)
+            {
+                mfxStatus mfxFindRes = session->m_pScheduler->FindOldestActiveSyncPoint(session->m_pDECODE.get(), &syncPoint);
+                // If we can't find any active SyncPoint's it means that next call to MFXVideoDECODE_DecodeFrameAsync should already accept a task.
+                // Let's repeat the call right here instead of making application do it.
+                if (MFX_ERR_NOT_FOUND == mfxFindRes)
```

Comment:
You have `do {...} while(false);` loop. So, calling `continue` won't make a trick: you will go to the next iteration, but loop condition will be checked first, so you will exit the loop instead of trying it once again.  I think you need:
```
while(1) {
    ...
    if (MFX_ERR_NOT_FOUND == mfxFindRes)
        continue;
    ...
    break;
}
```
kind of construction.